src/Forcings/forcing.jl

```diff
@@ -0,0 +1,93 @@
+the signature of `func` must include them. For example, if `field_dependencies=(:u, :S)`
+(and `parameters` are _not_ provided), then `func` must be callable with the signature
+
+    `func(x, y, z, t, u, S)`
+
+where `u` is assumed to be the `u`-velocity component, and `S` is a tracer. Note that any field
+which does not have the name `u`, `v`, or `w` is assumed to be a tracer and must be present
+in `model.tracers`.
+
+If `discrete_form=false` (the default) and `parameters` are provided, then the _last_ argument
+to `func` must be `parameters`. For example, if `func` has no `field_dependencies` but does
+depend on `parameters`, then it must be callable with the signature
+
+    `func(x, y, z, t, parameters)`
+
+With `field_dependencies=(:u, :v, :w, :c)` and `parameters`, then `func` must be
+callable with the signature
+
+    `func(x, y, z, t, u, v, w, c, parameters)`
+
+If `discrete_form=true` then `func` must be callable with the "discrete form"
+
+    `func(i, j, k, grid, clock, model_fields)`
+
+where `i, j, k` is the grid point at which the forcing is applied, `grid` is `model.grid`,
+`clock.time` is the current simulation time and `clock.iteration` is the current model iteration,
+and `model_fields` is a `NamedTuple` with `u, v, w`, the fields in `model.tracers`,
+and the fields in `model.diffusivities`, each of which is an `OffsetArray`s (or `NamedTuple`s
+of `OffsetArray`s depending on the turbulence closure) of field data.
+
+When `discrete_form=true` and `parameters` _is_ specified, `func` must be callable with the signature.
+
+    `func(i, j, k, grid, clock, model_fields, parameters)`
+    
+`parameters` is arbitrary in principle, however GPU compilation can place
+constraints on `typeof(parameters)`.
+
+Examples
+=======
+
+```julia
+# Parameterized forcing
+parameterized_func(x, y, z, t, p) = p.μ * exp(z / p.λ) * cos(p.ω * t)
+
+v_forcing = Forcing(parameterized_func, parameters = (μ=42, λ=0.1, ω=π))
+
+# Field-dependent forcing
+growth_in_sunlight(x, y, z, t, P) = exp(z) * P
+
+plankton_forcing = Forcing(growth_in_sunlight, field_dependencies=:P)
+
+# Parameterized, field-dependent forcing
+tracer_relaxation(x, y, z, t, c, p) = p.μ * exp((z + p.H) / p.λ) * (p.dCdz * z - c) 
+
+c_forcing = Forcing(tracer_relaxation,
+                    field_dependencies = :c,
+                            parameters = (μ=1/60, λ=10, H=1000, dCdz=1))
+
+# Unparameterized discrete-form forcing function
+filtered_relaxation(i, j, k, grid, clock, model_fields) =
+    @inbounds - (model_fields.c[i-1, j, k] + model_fields.c[i, j, k] + model_fields.c[i+1, j, k]) / 3
+
+simple_forcing = Forcing(simple_nonlinear_source, discrete_form=true)
+
+# Discrete-form forcing function with parameters
+masked_damping(i, j, k, grid, clock, model_fields, parameters) = 
+    @inbounds - parameters.μ * exp(grid.zC[k] / parameters.λ) * model_fields.u[i, j, k]
+
+masked_damping_forcing = Forcing(masked_damping, parameters=(μ=42, λ=π), discrete_form=true)
+```
```

Comment:
Probably needs a few `jldoctest` blocks.

See example at: https://github.com/CliMA/Oceananigans.jl/blob/e808a821d78949b0482b8372533e992cabb5d65f/src/OutputWriters/netcdf_output_writer.jl#L87-L157

Reply:
Fixed. Take a look and let me know what you think.